skfem/assembly/form/bilinear_form.py

```diff
@@ -42,6 +42,7 @@ class BilinearForm(Form):
     def assemble(self,
                  u: Basis,
                  v: Optional[Basis] = None,
+                 dtype=np.float,
```

Comment:
I wonder whether there's a type-hint for `dtype`.  I note that `isinstance(np.float, type)` so would it make sense to write
```python
dtype: type = np.float
```
?

(I don't know the answer here.)  

So far I don't think anything formal or automatic is being done with the type-hints in scikit-fem (except perhaps in users' IDEs?), but one day that might be useful; e.g. they might be exploited by optimizing compilers.

----

On looking further into this tangential question, I see that it's quite a topic. numpy/numpy#7370 numpy/numpy#16515

Reply:
Unfortunately I'm very new to Python, so I'm not familiar with many things yet :-). Had to google type hinting.

It does seem like a good idea to have dtype be of type 'type'. Though I found that there is a legacy convention, which allows defining the type as a string (for instance dtype='f') for numpy arrays, but I don't think that needs to be supported.

Reply:
Regarding the Orr-Sommerfeld example and complex systems. I was working on multiconductor transmission line computations, with a nonzero damping matrix from a resistive absorbing boundary condition. This results in a complex system, but with all assembled matrixes real. Thus condense and solve work fine with complex matrixes. It wasn't until I added a capacitive boundary condition, which made the damping matrix complex that I arrived to problems.